google-cloud-firestore/src/main/java/com/google/cloud/firestore/DocumentReference.java

```diff
@@ -372,41 +513,45 @@ public ApiFuture<DocumentSnapshot> get(FieldMask fieldMask) {
    */
   @Nonnull
   public Iterable<CollectionReference> listCollections() {
-    ListCollectionIdsRequest.Builder request = ListCollectionIdsRequest.newBuilder();
-    request.setParent(path.toString());
-    final ListCollectionIdsPagedResponse response;
-    try {
+    TraceUtil.Span span = getTraceUtil().startSpan(TraceUtil.SPAN_NAME_DOC_REF_LIST_COLLECTIONS);
+    try (Scope ignored = span.makeCurrent()) {
+      ListCollectionIdsRequest.Builder request = ListCollectionIdsRequest.newBuilder();
+      request.setParent(path.toString());
+      final ListCollectionIdsPagedResponse response;
       response =
           ApiExceptions.callAndTranslateApiException(
               rpcContext.sendRequest(
                   request.build(), rpcContext.getClient().listCollectionIdsPagedCallable()));
+      Iterable<CollectionReference> result =
+          new Iterable<CollectionReference>() {
+            @Override
+            @Nonnull
+            public Iterator<CollectionReference> iterator() {
+              final Iterator<String> iterator = response.iterateAll().iterator();
+              return new Iterator<CollectionReference>() {
+                @Override
+                public boolean hasNext() {
+                  return iterator.hasNext();
+                }
+
+                @Override
+                public CollectionReference next() {
+                  return DocumentReference.this.collection(iterator.next());
+                }
+
+                @Override
+                public void remove() {
+                  throw new UnsupportedOperationException("remove");
+                }
+              };
+            }
+          };
+      span.end();
+      return result;
     } catch (ApiException exception) {
+      span.end(exception);
       throw FirestoreException.forApiException(exception);
     }
```

Comment:
maybe this is the diff, but why is the order changing here (catching exceptions then returning vs a catch all the way at the end)?

Reply:
I found the way the code was originally written hard to read: the `try` block used to only encapsulate part of the logic. I decided to improve this a bit by moving the `return` statements into the `try` block.

This shouldn't change the behavior because the only exception that could be thrown by the return statement is an "out of memory" exception (because of the `new`). And in the case of OOM, whether the return statement is in the `try` block or after it, the behavior would be the same (because the function doesn't catch any OOM exceptions). On the plus side though, by moving the return statement into the try block, we can make the "SPAN_NAME_DOC_REF_LIST_COLLECTIONS span" nicely encapsulate all the cpu time spent in this function.